server/src/main/java/org/apache/kafka/server/RPCProducerIdManager.java

```diff
@@ -0,0 +1,191 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.server;
+
+import org.apache.kafka.clients.ClientResponse;
+import org.apache.kafka.common.message.AllocateProducerIdsRequestData;
+import org.apache.kafka.common.message.AllocateProducerIdsResponseData;
+import org.apache.kafka.common.protocol.Errors;
+import org.apache.kafka.common.requests.AllocateProducerIdsRequest;
+import org.apache.kafka.common.requests.AllocateProducerIdsResponse;
+import org.apache.kafka.common.utils.Time;
+import org.apache.kafka.server.common.ProducerIdsBlock;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+
+public class RPCProducerIdManager extends ProducerIdManager {
+
+    private static final Logger log = LoggerFactory.getLogger(RPCProducerIdManager.class);
+    private final String logPrefix;
+
+    private final int brokerId;
+    private final Time time;
+    private final Supplier<Long> brokerEpochSupplier;
+    private final NodeToControllerChannelManager controllerChannel;
+
+    // Visible for testing
+    public AtomicReference<ProducerIdsBlock> nextProducerIdBlock = new AtomicReference<>(null);
+    private final AtomicReference<ProducerIdsBlock> currentProducerIdBlock = new AtomicReference<>(ProducerIdsBlock.EMPTY);
+    private final AtomicBoolean requestInFlight = new AtomicBoolean(false);
+    private final AtomicLong backoffDeadlineMs = new AtomicLong(NO_RETRY);
+
+    public RPCProducerIdManager(int brokerId,
+                                Time time,
+                                Supplier<Long> brokerEpochSupplier,
+                                NodeToControllerChannelManager controllerChannel
+    ) {
+        this.brokerId = brokerId;
+        this.time = time;
+        this.brokerEpochSupplier = brokerEpochSupplier;
+        this.controllerChannel = controllerChannel;
+        this.logPrefix = "[RPC ProducerId Manager " + brokerId + "]: ";
+    }
+
+
+    @Override
+    public Long generateProducerId() throws Exception {
+        Long result = null;
```

Comment:
we can use `return` to simplify the code.
```java
        var iteration = 0;
        while (iteration <= ITERATION_LIMIT) {
            var claimNextId = currentProducerIdBlock.get().claimNextId();
            if (claimNextId.isPresent()) {
                long nextProducerId = claimNextId.get();
                // Check if we need to prefetch the next block
                long prefetchTarget = currentProducerIdBlock.get().firstProducerId() +
                        (long) (currentProducerIdBlock.get().size() * PID_PREFETCH_THRESHOLD);
                if (nextProducerId == prefetchTarget) {
                    maybePrefetchNextBlock();
                }
                return nextProducerId;
            } else {
                // Check the next block if current block is full
                var block = nextProducerIdBlock.getAndSet(null);
                if (block == null) {
                    // Return COORDINATOR_LOAD_IN_PROGRESS rather than REQUEST_TIMED_OUT since older clients treat the error as fatal
                    // when it should be retriable like COORDINATOR_LOAD_IN_PROGRESS.
                    maybeRequestNextBlock();
                    throw Errors.COORDINATOR_LOAD_IN_PROGRESS.exception("Producer ID block is full. Waiting for next block");
                } else {
                    currentProducerIdBlock.set(block);
                    requestInFlight.set(false);
                    iteration++;
                }
            }
        }
        throw Errors.COORDINATOR_LOAD_IN_PROGRESS.exception("Producer ID block is full. Waiting for next block");
```